lib/gssapi/spnego/init_sec_context.c

```diff
@@ -510,14 +502,37 @@ spnego_reply
 	if (GSS_ERROR(ret)) {
 	    HEIMDAL_MUTEX_unlock(&ctx->ctx_id_mutex);
 	    free_NegotiationToken(&resp);
+            /*
+             * If the acceptor rejected, preserve the minor status information
+             * from the mechanism, but indicate GSS_S_BAD_MECH.
+             */
+            if (resp.u.negTokenResp.negResult != NULL &&
+                *resp.u.negTokenResp.negResult == reject)
+                ret = GSS_S_BAD_MECH;
```

Comment:
Clobbering ret in this case gives different behavior to had the application invoked the mechanism directly. Suggest honoring mechanism return code.

Reply:
Also, you need to move free_NegotiationToken() after this

Reply:
But the acceptor at the SPNEGO layer said `reject`, meaning they rejected the _mechanism_.  This would only happen on the first round trip.  Any more legs and I think it's not sensible to reject the choice of mechanism -- you've already committed to it by then.

Reply:
Oy, right about the free.

Reply:
If the acceptor is rejecting the *choice* then it should omit the responseToken.

Reply:
Pushed a fix.  I'm leaving it as `GSS_S_BAD_MECH`.  If we have a counter of legs I could make it do that only on the second.